api/admin/controller/collection_settings.py

```diff
@@ -0,0 +1,266 @@
+                return COLLECTION_NAME_ALREADY_IN_USE
+            collection.create_external_integration(fields.get("protocol"))
+
+        collection.name = name
+        [protocol] = [p for p in self.protocols if p.get("name") == protocol]
+
+        settings = self.validate_parent(protocol, collection)
+        if isinstance(settings, ProblemDetail):
+            return settings
+        settings_error = self.process_settings(settings, collection)
+        if settings_error:
+            return settings_error
+
+        libraries_error = self.process_libraries(protocol, collection)
+        if libraries_error:
+            return libraries_error
+
+        if is_new:
+            return Response(unicode(collection.id), 201)
+        else:
+            return Response(unicode(collection.id), 200)
+
+    def validate_form_fields(self, is_new, **fields):
+        if not fields.get("name"):
+            return MISSING_COLLECTION_NAME
+        if fields.get("protocol") and fields.get("protocol") not in [p.get("name") for p in self.protocols]:
+            return UNKNOWN_PROTOCOL
+        if not fields.get("protocol"):
+            return NO_PROTOCOL_FOR_NEW_SERVICE
+        if "collection" in fields and not fields.get("collection"):
+            return MISSING_COLLECTION
+        if fields.get("collection"):
+            return self.validate_collection(**fields)
+
+    def validate_collection(self, **fields):
+        if fields.get("protocol") != fields.get("collection").protocol:
+            return CANNOT_CHANGE_PROTOCOL
+        if fields.get("name") != fields.get("collection").name:
+            collection_with_name = get_one(self._db, Collection, name=fields.get("name"))
+            if collection_with_name:
+                return COLLECTION_NAME_ALREADY_IN_USE
+
+    def validate_parent(self, protocol, collection):
+        parent_id = flask.request.form.get("parent_id")
+        if parent_id and not protocol.get("child_settings"):
+            self._db.rollback()
+            return PROTOCOL_DOES_NOT_SUPPORT_PARENTS
+
+        if parent_id:
+            parent = get_one(self._db, Collection, id=parent_id)
+            if not parent:
+                self._db.rollback()
+                return MISSING_PARENT
+            collection.parent = parent
+            settings = protocol.get("child_settings")
+        else:
+            collection.parent = None
+            settings = protocol.get("settings")
+
+        return settings
+
+    def validate_external_account_id_setting(self, value, setting):
+        if not value and not setting.get("optional"):
+            # Roll back any changes to the collection that have already been made.
+            self._db.rollback()
+            return INCOMPLETE_CONFIGURATION.detailed(
+                _("The collection configuration is missing a required setting: %(setting)s",
+                  setting=setting.get("label")))
+
+    def get_integration_id(self, value):
```

Comment:
Call this `get_mirror_integration_id` -- otherwise it sounds like a general helper method.